XML Improvements

Holzhaus opened the pull request.

Holzhaus committed:
refactor(xml): Add `xml` feature flag

Holzhaus committed:
feat(xml): Use proper `NaiveDate` format

Holzhaus committed:
refactor(xml): Remove `entries` and rename `tracks` in `Collection`

Holzhaus committed:
feat(xml): Make fields public

Holzhaus committed:
feat(xml): Parse ratings

Holzhaus committed:
feat(xml): Parse position mark colors

Swiftb0y commented on src/main.rs:
```diff
@@ -166,10 +166,12 @@ fn dump_setting(path: &PathBuf) -> rekordcrate::Result<()> {
     let file = std::fs::File::open(path)?;
     let reader = std::io::BufReader::new(file);
-    let document: Document = quick_xml::de::from_reader(reader).expect("failed to deserialize XML");
+    let document: rekordcrate::xml::Document =
+        quick_xml::de::from_reader(reader).expect("failed to deserialize XML");
```
have you figured out how to avoid the `expect` here yet? (the thing I criticized in the previous PR).

Swiftb0y commented on src/xml.rs:
```diff
@@ -27,221 +33,429 @@ pub struct Document {
+    pub tracks: Vec<Track>,
+}
+
+impl Serialize for Collection {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        #[derive(Serialize)]
+        struct Value<'a> {
+            /// Number of TRACK in COLLECTION
+            #[serde(rename = "@Count")]
+            entries: usize,
+            /// Tracks
+            #[serde(rename = "TRACK")]
+            tracks: &'a Vec<Track>,
+        }
+
+        let value = Value {
+            entries: self.tracks.len(),
+            tracks: &self.tracks,
+        };
+
+        value.serialize(serializer)
+    }
```
I feel like this could be simplified by using `#[serde(getter = "tracks.len()")]` instead?

Swiftb0y commented on src/xml.rs:
```diff
@@ -27,221 +33,429 @@ pub struct Document {
+        #[derive(Serialize)]
+        struct Value<'a> {
+            /// Number of TRACK in COLLECTION
+            #[serde(rename = "@Count")]
+            entries: usize,
```
The commit message doesn't explain why this is now `@Count` when it was previously `@Entries`. Can you elaborate?

Holzhaus replied:
The format is inconsistent, and it think I confused the item count for the Collection element ("Entries") with the item count for the Node element ("Count"). The original version was correct. I will revert.

<img width="1126" height="252" alt="image" src="https://github.com/user-attachments/assets/0200c8e2-1f7a-4f84-9c39-af49c0b62862" />

Swiftb0y commented on src/xml.rs:
```diff
@@ -27,221 +33,429 @@ pub struct Document {
+            Self::Turquoise => write!(f, "Turquoise"),
+            Self::Blue => write!(f, "Blue"),
+            Self::Violet => write!(f, "Violet"),
+            Self::Custom(rgb) => write!(f, "#{:02X}{:02X}{:02X}", rgb.r, rgb.g, rgb.b),
```
Alternative: delegate to the inner type if you don't need a stable representation 
https://docs.rs/rgb/latest/rgb/type.RGB8.html#impl-Display-for-Rgb%3CT%3E
```suggestion
            Self::Custom(rgb) => write!(f, "{}", rgb),
```

Swiftb0y commented on src/xml.rs:
```diff
@@ -27,221 +33,429 @@ pub struct Document {
-    Unknown(i32),
+#[derive(Debug, PartialEq, Clone, Copy, Serialize_repr, Deserialize_repr)]
+#[repr(u8)]
+pub enum StarRating {
+    /// Zero Stars
+    Zero = 0x00,
+    /// One Star
+    One = 0x33,
+    /// Two Stars
+    Two = 0x66,
+    /// Three Stars
+    Three = 0x99,
+    /// Four Stars
+    Four = 0xCC,
+    /// Five Stars
+    Five = 0xFF,
 }
```
since this is a `u8` under the hood, do fractional stars also exist (eg `230` corresponding to roughly `4.5` stars)? Do those need to be modeled here?

Holzhaus replied:
I think Rekordbox UI doesn't allow this. We could of course try to make a `Custom(u8)` variant, but I don't think it's needed. If it turns out to be necessary, we'll bump the version :)

Swiftb0y reviewed (commented):
large lgtm, couple thoughts

lazulit3 committed:
test(xml): add empty playlist case to XML test data

Updates the XML test data to include an empty playlist.

Holzhaus committed:
chore(main): Remove `expect()` from XML deserialization call

acrilique reviewed (approved):
This LGTM, may I take the freedom to fix the conflicts and merge?

acrilique committed:
fix(xml): do not error on empty playlist nodes

Fixes #145

When a playlist has zero tracks (Entries="0"), the map is empty after consuming
attributes, and those structs' Vec fields lacked #[serde(default)], causing
"missing field TRACK". The same bug existed for empty folders with Count="0".

acrilique commented:
Just rebased fixing all conflicts and added a fix and a test for #145 , thus superseding #146
I'll leave this here for a few days to give you all the change to review.

acrilique commented:
Merging this to keep moving forward despite the lack of feedback
